proofs/Settlement.thy

```diff
@@ -0,0 +1,160 @@
+"RmaxP1_pre T1 L1 T2 L2 D1 W1 =
+  impl_sub (impl_add (impl_sub (TLmax2 T2 L2) (TLmax1 T1 L1)) (Some D1)) (Some W1)"
+
+definition TAD :: "int \<Rightarrow> int \<Rightarrow> int \<Rightarrow> int \<Rightarrow> impl_number" where
+"TAD D1 D2 W1 W2 = impl_sub (impl_sub (impl_add (Some D1) (Some D2)) (Some W1)) (Some W2)"
+
+definition RmaxP1 :: "int \<Rightarrow> int \<Rightarrow> int \<Rightarrow> int \<Rightarrow> int \<Rightarrow> int \<Rightarrow> int \<Rightarrow> int \<Rightarrow> impl_number" where
+"RmaxP1 T1 L1 T2 L2 D1 W1 D2 W2 =
+  impl_min (TAD D1 D2 W1 W2) (RmaxP1_pre T1 L1 T2 L2 D1 W1)"
+
+definition SL2 :: "int \<Rightarrow> int \<Rightarrow> int \<Rightarrow> int \<Rightarrow> int \<Rightarrow> int \<Rightarrow> int \<Rightarrow> int \<Rightarrow> impl_number" where
+"SL2 T1 L1 T2 L2 D1 W1 D2 W2 =
+   impl_min (RmaxP1 T1 L1 T2 L2 D1 W1 D2 W2) (Some L2)"
+
+definition S1 :: "int \<Rightarrow> int \<Rightarrow> int \<Rightarrow> int \<Rightarrow> int \<Rightarrow> int \<Rightarrow> int \<Rightarrow> int \<Rightarrow> impl_number" where
+"S1 T1 L1 T2 L2 D1 W1 D2 W2 =
+   impl_sub (RmaxP1 T1 L1 T2 L2 D1 W1 D2 W2) (SL2 T1 L1 T2 L2 D1 W1 D2 W2)"
+
+definition RmaxP2 :: "int \<Rightarrow> int \<Rightarrow> int \<Rightarrow> int \<Rightarrow> int \<Rightarrow> int \<Rightarrow> int \<Rightarrow> int \<Rightarrow> impl_number"
+  where
+"RmaxP2 T1 L1 T2 L2 D1 W1 D2 W2 =
+    impl_sub (TAD D1 D2 W1 W2) (RmaxP1 T1 L1 T2 L2 D1 W1 D2 W2)"
+
+definition SL1 :: "int \<Rightarrow> int \<Rightarrow> int \<Rightarrow> int \<Rightarrow> int \<Rightarrow> int \<Rightarrow> int \<Rightarrow> int \<Rightarrow> impl_number"
+  where
+"SL1 T1 L1 T2 L2 D1 W1 D2 W2 = impl_min (RmaxP2 T1 L1 T2 L2 D1 W1 D2 W2) (Some L1)"
+
+definition S2 :: "int \<Rightarrow> int \<Rightarrow> int \<Rightarrow> int \<Rightarrow> int \<Rightarrow> int \<Rightarrow> int \<Rightarrow> int \<Rightarrow> impl_number"
+  where
+"S2 T1 L1 T2 L2 D1 W1 D2 W2
+ = impl_sub (RmaxP2 T1 L1 T2 L2 D1 W1 D2 W2) (SL1 T1 L1 T2 L2 D1 W1 D2 W2)"
+
+definition spec_s1 :: "int \<Rightarrow> int \<Rightarrow> int \<Rightarrow> int \<Rightarrow> int \<Rightarrow> int"
+  where
+"spec_s1 T1 T2 D1 W1 L1 = D1 - W1 + T2 - T1 - L1"
+
+definition spec_s2 :: "int \<Rightarrow> int \<Rightarrow> int \<Rightarrow> int \<Rightarrow> int \<Rightarrow> int"
+  where
+"spec_s2 T1 T2 D2 W2 L2 = D2 - W2 + T1 - T2 - L2"
+
+lemma s1_correct :
+"
+valid T1 \<Longrightarrow>
+valid T2 \<Longrightarrow>
+valid L1 \<Longrightarrow>
+valid L2 \<Longrightarrow>
+valid D1 \<Longrightarrow>
+valid D2 \<Longrightarrow>
+valid (D1 + D2) \<Longrightarrow> (* (12) *)
+valid W1 \<Longrightarrow>
+valid W2 \<Longrightarrow>
+valid (T1 + L1) \<Longrightarrow> (* 10 *)
+valid (T2 + L2) \<Longrightarrow> (* 10 *)
+L1 <= D1 - W1 + T2 - T1 \<Longrightarrow> (* (5) *)
+D1 - W1 + T2 - T1 - L1 \<le> D1 + D2 - W1 - W2 \<Longrightarrow> (* (5) *)
+-(D1 - W1) <= T2 + L2 - T1 - L1 \<Longrightarrow> (* (7) *)
```

Comment:
Hummm, the rule `( 7 )` from the spec does not seem to sustain itself on its own, I think it should be broken down into two.

The current statement is `(7 R) -(D1 - W1) <= T2 + L2 - T1 - L1 <= D2 - W2`, which boils down to two things (note that there is only one locked amount):

- A node cannot commit/send more then it has `-(D1 - W1) <= T2 - T1 - L1`
- A node must not accept from its partner more then he has `T2 + L2 - T1 <= D2 - W2`

Having both locked amounts in there would make this valid:

`W1=0`, `W2=0`, `T1=0`, `T2=0`, reduces `-(D1 - W1) <= T2 + L2 - T1 - L1 <= D2 - W2` to `-D1 <= L2 - L1 <= D2`

D1 | D2 | L1 | L2 | -D1 <= L2 - L1 <= D2
--- | --- | --- | --- | ---
1 | 1 | 0 | 0 | true
1 | 1 | 1 | 0 | true
1 | 1 | 2 | 1 | true

The last line above should not be true, however, with `-(D1 - W1) <= T2 - T1 - L1`

D1 | L1| -D1 <= - L1
--- | --- | ---
1 | 0 | true
1 | 1 | true
1 | 2 | false

I believe the above worked because of the stricter definition just above `L1 <= D1 - W1 + T2 - T1` (but I have not checked this hypothesis)

Reply:
Now I tried if Isabelle understands `0 <= 1 <= 2`; no it doesn't.

I guess
```
 -(D1 - W1) <= T2 + L2 - T1 - L1 /\ T2 + L2 - T1 - L1 <= D2 - W2 (7 R)
```
is the best approximation.  I'll use ^ unless I hear otherwise.

Reply:
The 3rd case seems weird:

D1 | D2 | L1 | L2 | -D1 <= L2 - L1 <= D2
--- | --- | --- | --- | ---
1 | 1 | 2 | 1 | true

because to get to that case, we need a state of:

D1 | D2 | L1 | L2
--- | --- | --- | --- 
1 | 1 | 1 | 1

Here, both available balances are `0`:  `AB1 = 1 - 0 + 0 - 0 - 1 = 0` & `AB2 = 1 - 0 + 0 - 0 - 1 = 0`. So, none of the participants can make transfers -> under `(5 R)` you can't have `L1 == 2`.
Am I missing something?

Reply:
If this discussion is about changing the condition in the spec, please take it there.

Reply:
updated my comment - I missed the point first time I read it.
So, just to conclude:
So `-(D1 - W1) <= T2 + L2 - T1 - L1 <= D2 - W2` is useful in the demonstration for `T2 + L2 - T1 - L1 <= D2 - W2` - this is used solve the `+ D1` overflow from `(T2 + L2) - (T1 + L1) + D1 - W1`.
Agreed that the first part `-(D1 - W1) <= T2 + L2 - T1 - L1` is less strict than needed to sustain itself. This was added after this discussion: https://github.com/raiden-network/spec/pull/119#discussion_r214018326.
They can be separated if needed.

Reply:
Yes, I dropped the first part and the proof still works.

Reply:
And I opened an issue about changing the spec https://github.com/raiden-network/spec/issues/209 .